jplag/src/main/java/de/jplag/GreedyStringTiling.java

```diff
@@ -4,100 +4,56 @@
-     */
-    public void createHashes(TokenList tokenList, int hashLength, boolean makeTable) {
-        // Here the upper boundary of the hash length is set.
-        // It is determined by the number of bits of the 'int' data type and the number of tokens.
-        if (hashLength < 1) {
-            hashLength = 1;
-        }
-        hashLength = (hashLength < 26 ? hashLength : 25);
-
-        if (tokenList.size() < hashLength) {
-            return;
-        }
-
-        int modulo = ((1 << 6) - 1);   // Modulo 64!
-
-        int loops = tokenList.size() - hashLength;
-        tokenList.tokenHashes = (makeTable ? new TokenHashMap(3 * loops) : null);
-        int hash = 0;
-        int hashedLength = 0;
-        for (int i = 0; i < hashLength; i++) {
-            hash = (2 * hash) + (tokenList.getToken(i).type & modulo);
-            hashedLength++;
-            if (tokenList.getToken(i).isMarked()) {
-                hashedLength = 0;
-            }
-        }
-        int factor = (hashLength != 1 ? (2 << (hashLength - 2)) : 1);
-
-        if (makeTable) {
-            for (int i = 0; i < loops; i++) {
-                if (hashedLength >= hashLength) {
-                    tokenList.getToken(i).setHash(hash);
-                    tokenList.tokenHashes.put(hash, i);   // add into hashtable
-                } else {
-                    tokenList.getToken(i).setHash(-1);
-                }
-                hash -= factor * (tokenList.getToken(i).type & modulo);
-                hash = (2 * hash) + (tokenList.getToken(i + hashLength).type & modulo);
-                if (tokenList.getToken(i + hashLength).isMarked()) {
-                    hashedLength = 0;
-                } else {
-                    hashedLength++;
-                }
-            }
-        } else {
-            for (int i = 0; i < loops; i++) {
-                tokenList.getToken(i).setHash((hashedLength >= hashLength) ? hash : -1);
-                hash -= factor * (tokenList.getToken(i).type & modulo);
-                hash = (2 * hash) + (tokenList.getToken(i + hashLength).type & modulo);
-                if (tokenList.getToken(i + hashLength).isMarked()) {
-                    hashedLength = 0;
-                } else {
-                    hashedLength++;
-                }
-            }
-        }
-        tokenList.hashLength = hashLength;
+        this.minimumMatchLength = options.minimumTokenMatch();
     }
 
     public final JPlagComparison compare(Submission firstSubmission, Submission secondSubmission) {
-        return swapAndCompare(firstSubmission, secondSubmission, false);
+        return swapAndCompare(firstSubmission, secondSubmission);
     }
 
-    public final JPlagComparison compareWithBaseCode(Submission firstSubmission, Submission secondSubmission) {
-        return swapAndCompare(firstSubmission, secondSubmission, true);
+    public final JPlagComparison compareWithBaseCode(Submission submission, Submission baseCodeSubmission) {
```

Comment:
This change is confusing me a bit. Previously, there were two methods. One for comparison without factoring in the basecode and one for comparison where basecode matches were ignored. The second one was required for the main comparison, whereas the first one was required to calculate matching sections with the basecode for each submission in firsthand. Does it now mean after the refactoring that `compareWithBaseCode()` is the first one and `compare()` is the the second one?

Reply:
I would rather say that the previous API was confusing 😄.
In the old state, the `isBaseCodeMarking` parameter was used to identify that a base code comparison is taking place and thus the token hashes are computed without considering the base code (which is only needed as the base code submission itself is tagged as base code and this has to be ignored) and then the matches are used to identify what was a base code. Otherwise, the hashes for subsequences containing a baseCode marked token were set to `-1` to avoid matches. So there was actually never the intention to "ignore" the base code markings but rather only bad separation of concerns which integrated marking of base code tokens into the core algorithm.

With the new approach, the base code submission tokens are never tagged as base code such that their subsequence hashes are always computed correctly, and the base code markings are created outside of the core algorithm.

Besides that, the `compare` and `compareWithBasecode` have the same semantics as the ones in the old version. But I will add documentation to the methods.

Reply:
Ah okay, now I get it. That makes sense! +1 for the documentation.